sapicore/data/__init__.py

```diff
@@ -0,0 +1,474 @@
+
+        This method is meant to be implemented by child classes that deal with arbitrarily formatted external sets.
+        It may also serve as a bridge between Sapicore and specialized libraries, e.g. MNE for EEG/MEG data.
+
+        Warning
+        -------
+        The resulting file may follow any convention the user would like to adopt. Sapicore makes no assumptions;
+        it is the user's responsibility to implement this method and :meth:`access_data` below in a compatible way.
+
+        """
+        pass
+
+    def access_data(self, index: int | list[int] | tuple[int] | Tensor, axis: int = None):
+        """Specifies how to access data by mapping indices to actual samples (e.g., from file(s) in `root`).
+
+        The default implementation slices into `self.samples` to accommodate the trivial cases where the user has
+        directly initialized this :class:`Data` object with a `samples` tensor or loaded its values by reading
+        a file that fits in memory (the latter case would be handled by :meth:`load`).
+
+        More sophisticated use cases may require lazy loading or navigating HDF5 files. That kind of logic should
+        be implemented here by derivative classes.
+
+        Parameters
+        ----------
+        index: int or tuple of int or list of int or Tensor
+            Index(es) to slice into.
+
+        axis: int, optional
+            Optionally, a specific axis along which to apply index selection.
+
+        Note
+        ----
+        Where audio/image files are concerned (each being a labeled "sample"), use this method to read
+        and potentially transform them, returning the finished product.
+
+        """
+        if axis:
+            return self.samples.index_select(axis, torch.as_tensor(index))
+        else:
+            return self.samples[index]
+
+    def add_descriptors(
+        self, file: str = None, include: list[str] = None, *args: AxisDescriptor, **kwargs: AxisDescriptor
+    ):
+        """Extracts labels from file(s) in the data `root` directory, if given, and adds named
+        :class:`AxisDescriptor` objects to this :class:`Data` set.
+
+        This default implementation is intended to capture the general case where descriptor variables are
+        given in the header of a tabular CSV and values in each row describe the samples (rows) of the
+        associated :class:`Data` object. User may specify which column names to include.
+
+        Parameters
+        ----------
+        file: str
+            Path to the file containing label information.
+
+        include: list of str
+            List of keys to generate label vectors from (e.g., column names in a CSV).
+
+        kwargs: AxisDescriptor
+            Directly add one or more named :class:`AxisDescriptor` to this :class:`Data` object.
+
+        """
+        # add labels from file if given, wrapping them in a descriptor object.
+        if file and os.path.exists(file):
```

Comment:
Absolutely right--adding descriptors from file could _maybe_ belong in `DriftDataset` where we rigidly decide how to _standardize the set (since it's tiny, it made sense to parse and dump the labels to a CSV, but that's not an assumption I'd make in general). I'll move it there for now and keep this one about the args and kwargs (directly add AxisDescriptor objects). Note how there's some uncertainty in the docstring--this was definitely a temporary solution I wanted to circle back to.